// discord-interactions/dist/webhooks.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebhookEventType = exports.WebhookType = void 0;
/**
 * @see {@link https://discord.com/developers/docs/events/webhook-events#webhook-types}
 */
var WebhookType;
(function (WebhookType) {
    /**
     * A ping.
     */
    WebhookType[WebhookType["PING"] = 0] = "PING";
    /**
     * A webhook event.
     */
    WebhookType[WebhookType["EVENT"] = 1] = "EVENT";
})(WebhookType || (exports.WebhookType = WebhookType = {}));
/**
 * Events to which an app can subscribe to
 * @see {@link https://discord.com/developers/docs/events/webhook-events#event-types}
 */
var WebhookEventType;
(function (WebhookEventType) {
    /**
     * Event sent when an app was authorized by a user to a server or their account.
     */
    WebhookEventType["APPLICATION_AUTHORIZED"] = "APPLICATION_AUTHORIZED";
    /**
     * Event sent when an entitlement was created.
     */
    WebhookEventType["ENTITLEMENT_CREATE"] = "ENTITLEMENT_CREATE";
    /**
     * Event sent when an user was added to a Quest.
     */
    WebhookEventType["QUEST_USER_ENROLLMENT"] = "QUEST_USER_ENROLLMENT";
})(WebhookEventType || (exports.WebhookEventType = WebhookEventType = {}));
